File: examples/run_fandango.py
import os
from typing import Any
from fandango import Fandango


class FandangoWrapper:
    def __init__(self, fan_file: str, kwargs: dict[str, Any]):
        with open(fan_file) as f:
            self.fan = Fandango(f, **kwargs)
        self.generator = self.fan.generate_solutions()


def setup(fan_file: str, kwargs: dict[str, Any]) -> FandangoWrapper:
    return FandangoWrapper(fan_file, kwargs)


def next_input(wrapper: FandangoWrapper) -> bytes:
    return bytes(next(wrapper.generator))


def parse_input(wrapper: FandangoWrapper, input: bytes) -> int:
    return len(list(wrapper.fan.parse(input)))


if __name__ == "__main__":
    # path relative to this script
    fan_file = os.path.dirname(__file__) + "/even_numbers.fan"
    gen = setup(fan_file, {})
    for i in range(10):
        input = next_input(gen)
        print(input)
        assert parse_input(gen, input) > 0
    assert parse_input(gen, b"1") == 0, parse_input(gen, b"1")
    assert parse_input(gen, b"0") == 1, parse_input(gen, b"0")
    assert parse_input(gen, b"2") == 1, parse_input(gen, b"2")

File: examples/run_fandango.rs
use clap::Parser;
use libafl_fandango_pyo3::fandango::{
    FandangoClient as _, FandangoInprocessModule, FandangoModuleInitError,
};

#[derive(Parser)]
#[command(name = "run_fandango")]
#[command(about = "Run the fandango interface in Python")]
struct Args {
    #[arg(short, long, default_value = "examples/even_numbers.fan")]
    fandango_file: String,
}

fn main() {
    if let Err(err) = run() {
        eprintln!("{err}");
        std::process::exit(1);
    }
}

fn run() -> Result<(), String> {
    let args = Args::parse();

    let mut fandango = match FandangoInprocessModule::new(&args.fandango_file, &[]) {
        Ok(fandango) => fandango,
        Err(FandangoModuleInitError::ModuleNotFoundError(e, tb)) => {
            return Err(format!(
                "A required Python module was not found. You may need to set the PYTHONPATH environment variable to the path of the Python interpreter, e.g. `export PYTHONPATH=$(echo .venv/lib/python*/site-packages)`. Underlying error:\n{}\n{}",
                e, tb
            ));
        }
        Err(e) => {
            return Err(format!("Error: {:?}", e));
        }
    };

    for _ in 0..10 {
        let input = fandango.next_input().unwrap();
        let num_parses = fandango.parse_input(&input).unwrap();
        println!(
            "{} can be parsed {} different ways", // this grammar always produces inputs that can be parsed in exactly one way — but this isn't necessarily true for all grammars
            String::from_utf8(input).unwrap(),
            num_parses
        );
        assert!(num_parses == 1);
    }

    Ok(())
}
